### PR TITLE
Generate pretty-formatted JSON cassettes

### PR DESCRIPTION
**Pull request by mainameiz**
YAML-serializer produces pretty-formatted output. And I think it is more useful/human-readable/understandable.
So I would like to mirror this behaviour in JSON-serializer.

As I see there is another option which will not change current behaviour of JSON-serializer - make a separate serializer which will make a pretty-formatted output (e.g. `Serializers::PrettyJSON`).

### CODE REVIEW

File: lib/vcr/cassette/serializers/json.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 module VCR
   class Cassette
     class Serializers
-      # The JSON serializer. Uses `MultiJson` under the covers.
```
Comment (mainameiz) [inline lib/vcr/cassette/serializers/json.rb:6]: This is no longer true. See https://github.com/vcr/vcr/pull/815.

Comment (olleolleolle) [reply lib/vcr/cassette/serializers/json.rb:6]: We would accept a tiny PR about removing it!

Comment (mainameiz) [reply lib/vcr/cassette/serializers/json.rb]: 👌 https://github.com/vcr/vcr/pull/841

**Comment by olleolleolle** [comment]
First off, thanks a lot for taking all this time to make the output more human-readable!

**Review by olleolleolle — DISMISSED** [review]
Thanks for adding this change.

**Comment by krainboltgreene** [comment]
Okay so like on a high level I'm on board with this, but this has got to be a performance hit, yes?

**Comment by mainameiz** [comment]
@krainboltgreene Hi!
Yes, there is some performance impact.

I made some tests...


<details>
<summary>benchmark.rb</summary>

```ruby
require 'json'
require 'benchmark/ips'

# Just three copies of the same cassette from tests.
cassete1 = {"http_interactions"=>[{"request"=>{"method"=>"get", "uri"=>"http://localhost:33781/", "body"=>{"encoding"=>"US-ASCII", "base64_string"=>""}, "headers"=>{"Accept-Encoding"=>["gzip;q=1.0,deflate;q=0.6,identity;q=0.3"], "Accept"=>["*/*"], "User-Agent"=>["Ruby"]}}, "response"=>{"status"=>{"code"=>200, "message"=>"OK "}, "headers"=>{"Content-Type"=>["text/html;charset=utf-8"], "Content-Length"=>["11"], "Server"=>["WEBrick/1.3.1 (Ruby/2.4.10/2020-03-31)"], "Date"=>["Thu, 26 Nov 2020 09:47:20 GMT"], "Connection"=>["Keep-Alive"]}, "body"=>{"encoding"=>"US-ASCII", "base64_string"=>"SGVsbG8gV29ybGQ=\n"}}, "recorded_at"=>"Thu, 26 Nov 2020 09:47:20 GMT"}], "recorded_with"=>"VCR 6.0.0"}
cassete2 = {"http_interactions"=>[{"request"=>{"method"=>"get", "uri"=>"http://localhost:33781/", "body"=>{"encoding"=>"US-ASCII", "base64_string"=>""}, "headers"=>{"Accept-Encoding"=>["gzip;q=1.0,deflate;q=0.6,identity;q=0.3"], "Accept"=>["*/*"], "User-Agent"=>["Ruby"]}}, "response"=>{"status"=>{"code"=>200, "message"=>"OK "}, "headers"=>{"Content-Type"=>["text/html;charset=utf-8"], "Content-Length"=>["11"], "Server"=>["WEBrick/1.3.1 (Ruby/2.4.10/2020-03-31)"], "Date"=>["Thu, 26 Nov 2020 09:47:20 GMT"], "Connection"=>["Keep-Alive"]}, "body"=>{"encoding"=>"US-ASCII", "base64_string"=>"SGVsbG8gV29ybGQ=\n"}}, "recorded_at"=>"Thu, 26 Nov 2020 09:47:20 GMT"}], "recorded_with"=>"VCR 6.0.0"}
cassete3 = {"http_interactions"=>[{"request"=>{"method"=>"get", "uri"=>"http://localhost:33781/", "body"=>{"encoding"=>"US-ASCII", "base64_string"=>""}, "headers"=>{"Accept-Encoding"=>["gzip;q=1.0,deflate;q=0.6,identity;q=0.3"], "Accept"=>["*/*"], "User-Agent"=>["Ruby"]}}, "response"=>{"status"=>{"code"=>200, "message"=>"OK "}, "headers"=>{"Content-Type"=>["text/html;charset=utf-8"], "Content-Length"=>["11"], "Server"=>["WEBrick/1.3.1 (Ruby/2.4.10/2020-03-31)"], "Date"=>["Thu, 26 Nov 2020 09:47:20 GMT"], "Connection"=>["Keep-Alive"]}, "body"=>{"encoding"=>"US-ASCII", "base64_string"=>"SGVsbG8gV29ybGQ=\n"}}, "recorded_at"=>"Thu, 26 Nov 2020 09:47:20 GMT"}], "recorded_with"=>"VCR 6.0.0"}

json = [cassete1, cassete2, cassete3]

Benchmark.ips do |x|
  x.report("generate") { |n| n.times { JSON.generate(json) } }
  x.report("pretty_generate") { |n| n.times { JSON.pretty_generate(json) } }

  x.compare!
end
```

</details>

<details>
<summary>ruby 2.4.10 (1.42x slower)</summary>

```
Warming up --------------------------------------
            generate     8.737k i/100ms
     pretty_generate     6.180k i/100ms
Calculating -------------------------------------
            generate     87.552k (± 0.6%) i/s -    445.587k in   5.089546s
     pretty_generate     61.632k (± 0.6%) i/s -    309.000k in   5.013809s

Comparison:
            generate:    87552.5 i/s
     pretty_generate:    61632.3 i/s - 1.42x  (± 0.00) slower
```

</details>

<details>
<summary>ruby 2.5.3 (1.16x slower)</summary>

```
Warming up --------------------------------------
            generate     3.087k i/100ms
     pretty_generate     2.581k i/100ms
Calculating -------------------------------------
            generate     30.782k (± 0.4%) i/s -    154.350k in   5.014388s
     pretty_generate     26.598k (± 0.8%) i/s -    134.212k in   5.046157s

Comparison:
            generate:    30781.9 i/s
     pretty_generate:    26598.4 i/s - 1.16x  (± 0.00) slower
```

</details>

<details>
<summary>ruby 2.6.6 (1.14x slower)</summary>

```
Warming up --------------------------------------
            generate     3.074k i/100ms
     pretty_generate     2.669k i/100ms
Calculating -------------------------------------
            generate     30.846k (± 0.3%) i/s -    156.774k in   5.082546s
     pretty_generate     26.952k (± 0.2%) i/s -    136.119k in   5.050477s

Comparison:
            generate:    30845.9 i/s
     pretty_generate:    26951.9 i/s - 1.14x  (± 0.00) slower
```

</details>

<details>
<summary>ruby 2.7.2 (1.39x slower)</summary>

```
Warming up --------------------------------------
            generate     8.759k i/100ms
     pretty_generate     6.323k i/100ms
Calculating -------------------------------------
            generate     87.489k (± 0.4%) i/s -    437.950k in   5.005851s
     pretty_generate     62.961k (± 0.5%) i/s -    316.150k in   5.021487s

Comparison:
            generate:    87489.2 i/s
     pretty_generate:    62961.0 i/s - 1.39x  (± 0.00) slower
```

</details>

<details>
<summary>ruby 3.0.0-preview1 (1.44x slower)</summary>

```
Warming up --------------------------------------
            generate     9.299k i/100ms
     pretty_generate     6.460k i/100ms
Calculating -------------------------------------
            generate     92.700k (± 0.5%) i/s -    464.950k in   5.015781s
     pretty_generate     64.303k (± 0.4%) i/s -    323.000k in   5.023167s

Comparison:
            generate:    92699.7 i/s
     pretty_generate:    64303.3 i/s - 1.44x  (± 0.00) slower
```

</details>


I'm open to implement another solution.

**Comment by krainboltgreene** [comment]
Okay so hilariously, based on this, Ruby 3 and 2.7 is just as slow as Ruby 2.4, but 3 times faster than 2.6?

Okay so there are two options in my mind:

  1. Merge it in, performance be damned.
  2. Allow for someone to switch-on the "unpretty" version (for performance).

I'll let you decide, @mainameiz

**Comment by mainameiz** [comment]
I suppose that overall performance is still very high:
- 61632 items per/second for ruby 2.4
- 26598 items per/second for ruby 2.5
- 26952 items per/second for ruby 2.6
- 62961 items per/second for ruby 2.7
- 62961 items per/second for ruby 3.0.0-preview1

I don't think that anybody needed to generate such many cassettes per second :)

I suppose, that VCR should be used to generate a cassette one time and then this cassette should beused in subsequent test runs.

**Comment by mainameiz** [comment]
@krainboltgreene WDYT?

**Review by olleolleolle — APPROVED** [review]
Looks good to me!



:hearts: :yellow_heart: :green_heart: :blue_heart: :purple_heart:

**Review by krainboltgreene — APPROVED** [review]

**Pull request merged**